File: src/main/java/ru/org/linux/topic/AddTopicController.java
/*
 * Copyright 1998-2022 Linux.org.ru
 *    Licensed under the Apache License, Version 2.0 (the "License");
 *    you may not use this file except in compliance with the License.
 *    You may obtain a copy of the License at
 *
 *        http://www.apache.org/licenses/LICENSE-2.0
 *
 *    Unless required by applicable law or agreed to in writing, software
 *    distributed under the License is distributed on an "AS IS" BASIS,
 *    WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
 *    See the License for the specific language governing permissions and
 *    limitations under the License.
 */

package ru.org.linux.topic;

import akka.actor.ActorRef;
import com.google.common.base.Strings;
import com.google.common.collect.ImmutableMap;
import org.apache.commons.io.IOUtils;
import org.springframework.beans.factory.annotation.Autowired;
import org.springframework.beans.factory.annotation.Qualifier;
import org.springframework.stereotype.Controller;
import org.springframework.validation.BindingResult;
import org.springframework.web.bind.WebDataBinder;
import org.springframework.web.bind.annotation.*;
import org.springframework.web.servlet.ModelAndView;
import org.springframework.web.servlet.view.RedirectView;
import org.springframework.web.util.UriComponentsBuilder;
import ru.org.linux.auth.*;
import ru.org.linux.csrf.CSRFNoAuto;
import ru.org.linux.csrf.CSRFProtectionService;
import ru.org.linux.gallery.Image;
import ru.org.linux.gallery.ImageService;
import ru.org.linux.gallery.UploadedImagePreview;
import ru.org.linux.group.Group;
import ru.org.linux.group.GroupDao;
import ru.org.linux.group.GroupPermissionService;
import ru.org.linux.markup.MarkupPermissions;
import ru.org.linux.markup.MessageTextService;
import ru.org.linux.poll.Poll;
import ru.org.linux.poll.PollVariant;
import ru.org.linux.realtime.RealtimeEventHub;
import ru.org.linux.search.SearchQueueSender;
import ru.org.linux.section.Section;
import ru.org.linux.section.SectionService;
import ru.org.linux.site.Template;
import ru.org.linux.markup.MarkupType;
import ru.org.linux.spring.dao.MessageText;
import ru.org.linux.tag.TagName;
import ru.org.linux.tag.TagService;
import ru.org.linux.user.User;
import ru.org.linux.user.UserErrorException;
import ru.org.linux.user.UserPropertyEditor;
import ru.org.linux.user.UserService;
import ru.org.linux.util.ExceptionBindingErrorProcessor;
import ru.org.linux.util.markdown.MarkdownFormatter;
import scala.Tuple2;

import javax.annotation.Nullable;
import javax.servlet.http.HttpServletRequest;
import javax.servlet.http.HttpSession;
import javax.validation.Valid;
import java.beans.PropertyEditorSupport;
import java.io.File;
import java.io.IOException;
import java.net.URL;
import java.nio.charset.StandardCharsets;
import java.util.*;

@Controller
public class AddTopicController {
  private final SearchQueueSender searchQueueSender;

  private final CaptchaService captcha;

  private FloodProtector dupeProtector;
  private IPBlockDao ipBlockDao;
  private final GroupDao groupDao;
  private final SectionService sectionService;

  private final TagService tagService;

  private final UserService userService;

  private final TopicPrepareService prepareService;

  private final GroupPermissionService groupPermissionService;

  private final AddTopicRequestValidator addTopicRequestValidator;

  private final ImageService imageService;

  private final TopicService topicService;

  private final ActorRef realtimeHubWS;
  private final MarkdownFormatter renderService;

  private static final int MAX_MESSAGE_LENGTH_ANONYMOUS = 8196;
  private static final int MAX_MESSAGE_LENGTH = 65536;

  public AddTopicController(SearchQueueSender searchQueueSender, CaptchaService captcha, SectionService sectionService,
                            TagService tagService, UserService userService, TopicPrepareService prepareService,
                            GroupPermissionService groupPermissionService,
                            AddTopicRequestValidator addTopicRequestValidator, ImageService imageService,
                            TopicService topicService, @Qualifier("realtimeHubWS") ActorRef realtimeHubWS,
                            MarkdownFormatter renderService, GroupDao groupDao) {
    this.searchQueueSender = searchQueueSender;
    this.captcha = captcha;
    this.sectionService = sectionService;
    this.tagService = tagService;
    this.userService = userService;
    this.prepareService = prepareService;
    this.groupPermissionService = groupPermissionService;
    this.addTopicRequestValidator = addTopicRequestValidator;
    this.imageService = imageService;
    this.topicService = topicService;
    this.realtimeHubWS = realtimeHubWS;
    this.renderService = renderService;
    this.groupDao = groupDao;
  }

  @Autowired
  public void setDupeProtector(FloodProtector dupeProtector) {
    this.dupeProtector = dupeProtector;
  }

  @Autowired
  public void setIpBlockDao(IPBlockDao ipBlockDao) {
    this.ipBlockDao = ipBlockDao;
  }

  @ModelAttribute("ipBlockInfo")
  private IPBlockInfo loadIPBlock(HttpServletRequest request) {
    return ipBlockDao.getBlockInfo(request.getRemoteAddr());
  }

  @RequestMapping(value = "/add.jsp", method = RequestMethod.GET)
  public ModelAndView add(@Valid @ModelAttribute("form") AddTopicRequest form, HttpServletRequest request) {
    Template tmpl = Template.getTemplate();

    if (form.getMode()==null) {
      form.setMode(tmpl.getFormatMode());
    }

    Map<String, Object> params = new HashMap<>(prepareModel(form.getGroup(), AuthUtil.getCurrentUser(), form.getGroup().getSectionId(), request));

    Group group = form.getGroup();

    if (tmpl.isSessionAuthorized() && !groupPermissionService.isTopicPostingAllowed(group, AuthUtil.getCurrentUser())) {
      ModelAndView errorView = new ModelAndView("errors/good-penguin");
      errorView.addObject("msgHeader", "Недостаточно прав для постинга тем в эту группу");
      errorView.addObject("msgMessage", groupPermissionService.getPostScoreInfo(group));

      return errorView;
    }

    return new ModelAndView("add", params);
  }

  private ImmutableMap<String, Object> prepareModel(@Nullable Group group, @Nullable User currentUser, int sectionId,
                                                    HttpServletRequest request) {
    ImmutableMap.Builder<String, Object> params = ImmutableMap.builder();

    Section section = sectionService.getSection(sectionId);

    try {
      URL helpResource = request.getServletContext().getResource("/help/new-topic-" + Section.getUrlName(sectionId) + ".md");
      if (helpResource != null) {
        String helpRawText = IOUtils.toString(helpResource, StandardCharsets.UTF_8);

        String addInfo = renderService.renderToHtml(helpRawText, false);

        params.put("addportal", addInfo);
      }
    } catch (IOException e) {
      throw new RuntimeException(e);
    }

    params.put("sectionId", sectionId);
    params.put("section", section);

    if (group!=null) {
      params.put("group", group);
      params.put("postscoreInfo", groupPermissionService.getPostScoreInfo(group));
      params.put("showAllowAnonymous", groupPermissionService.enableAllowAnonymousCheckbox(group, currentUser));

      params.put("imagepost", groupPermissionService.isImagePostingAllowed(section, currentUser));
    }

    return params.build();
  }

  private User postingUser(Template tmpl, AddTopicRequest form) {
    User user;

    if (!tmpl.isSessionAuthorized()) {
      if (form.getNick() != null) {
        user = form.getNick();
      } else {
        user = userService.getAnonymous();
      }
    } else {
      user = AuthUtil.getCurrentUser();
    }

    return user;
  }

  @RequestMapping(value="/add.jsp", method=RequestMethod.POST)
  @CSRFNoAuto
  public ModelAndView doAdd(
          HttpServletRequest request,
          @Valid @ModelAttribute("form") AddTopicRequest form,
          BindingResult errors,
          @ModelAttribute("ipBlockInfo") IPBlockInfo ipBlockInfo
  ) {
    Template tmpl = Template.getTemplate();
    HttpSession session = request.getSession();

    Group group = form.getGroup();

    Map<String, Object> params = new HashMap<>(prepareModel(form.getGroup(), AuthUtil.getCurrentUser(), form.getGroup().getSectionId(), request));

    Section section = null;

    if (group!=null) {
      section = sectionService.getSection(group.getSectionId());
    }

    User user = postingUser(tmpl, form);

    user.checkBlocked(errors);
    user.checkFrozen(errors);

    IPBlockDao.checkBlockIP(ipBlockInfo, errors, user);

    if (group!=null && !groupPermissionService.isTopicPostingAllowed(group, user)) {
      errors.reject(null, "Недостаточно прав для постинга тем в эту группу");
    }

    if (form.getMode()==null) {
      form.setMode(tmpl.getFormatMode());
    }

    if (group!=null && !groupPermissionService.enableAllowAnonymousCheckbox(group, user)) {
      form.setAllowAnonymous(true);
    }

    if (MarkupPermissions.allowedFormatsJava(AuthUtil.getCurrentUser()).stream().map(MarkupType::formId).noneMatch(s -> s.equals(form.getMode()))) {
      errors.rejectValue("mode", null, "Некорректный режим разметки");
      form.setMode(MarkupType.Lorcode$.MODULE$.formId());
    }

    MessageText message = MessageTextService.processPostingText(Strings.nullToEmpty(form.getMsg()), form.getMode());

    if (user.isAnonymous()) {
      if (message.text().length() > MAX_MESSAGE_LENGTH_ANONYMOUS) {
        errors.rejectValue("msg", null, "Слишком большое сообщение");
      }
    } else {
      if (message.text().length() > MAX_MESSAGE_LENGTH) {
        errors.rejectValue("msg", null, "Слишком большое сообщение");
      }
    }

    UploadedImagePreview imagePreview = null;

    if (section!=null && groupPermissionService.isImagePostingAllowed(section, user)) {
      if (groupPermissionService.isTopicPostingAllowed(group, user)) {
        File image = imageService.processUploadImage(request);

        imagePreview = imageService.processUpload(user, session, image, errors);
      }

      if (section.isImagepost() && imagePreview == null && !errors.hasErrors()) {
        errors.reject(null, "Изображение отсутствует");
      }
    }

    Poll poll = null;
    
    if (section!=null && section.isPollPostAllowed()) {
      poll = preparePollPreview(form);
    }

    Topic previewMsg = null;

    if (group!=null) {
      previewMsg = Topic.fromAddRequest(form, user, request.getRemoteAddr());

      Image imageObject = null;

      if (imagePreview!=null) {
        imageObject = new Image(0, 0, "gallery/preview/" + imagePreview.mainFile().getName());
      }

      List<String> tagNames = TagName.parseAndSanitizeTags(form.getTags());

      if (!groupPermissionService.canCreateTag(section, user)) {
        List<String> newTags = tagService.getNewTags(tagNames);

        if (!newTags.isEmpty()) {
          errors.rejectValue("tags", null, "Вы не можете создавать новые теги ("+ TagService.tagsToString(newTags)+")");
        }
      }

      PreparedTopic preparedTopic = prepareService.prepareTopicPreview(
              previewMsg,
              TagService.namesToRefs(tagNames),
              poll,
              message,
              imageObject
      );

      params.put("message", preparedTopic);

      TopicMenu topicMenu = prepareService.getTopicMenu(
              preparedTopic,
              AuthUtil.getCurrentUser(),
              tmpl.getProf(),
              true
      );

      params.put("topicMenu", topicMenu);
    }

    if (!form.isPreviewMode() && !errors.hasErrors()) {
      CSRFProtectionService.checkCSRF(request, errors);
    }

    if (!form.isPreviewMode() && !errors.hasErrors() && !tmpl.isSessionAuthorized()
      || ipBlockInfo.isCaptchaRequired()) {
      captcha.checkCaptcha(request, errors);
    }

    if (!form.isPreviewMode() && !errors.hasErrors()) {
      dupeProtector.checkDuplication(FloodProtector.Action.ADD_TOPIC, request.getRemoteAddr(), user.getScore() >= 100, errors);
    }

    if (!form.isPreviewMode() && !errors.hasErrors() && group != null) {
      return createNewTopic(request, form, session, group, params, section, user, message, imagePreview, previewMsg);
    } else {
      return new ModelAndView("add", params);
    }
  }

  private ModelAndView createNewTopic(HttpServletRequest request, AddTopicRequest form, HttpSession session,
                                      Group group, Map<String, Object> params, Section section, User user,
                                      MessageText message, UploadedImagePreview scrn, Topic previewMsg) {
    session.removeAttribute("image");

    Tuple2<Integer, Set<Integer>> result = topicService.addMessage(
            request,
            form,
            message,
            group,
            user,
            scrn,
            previewMsg
    );

    int msgid = result._1;

    if (!previewMsg.isDraft())  {
      searchQueueSender.updateMessageOnly(msgid);
      RealtimeEventHub.notifyEvents(realtimeHubWS, result._2);
    }

    String topicUrl = previewMsg.withId(msgid).getLink();

    if (!section.isPremoderated() || previewMsg.isDraft()) {
      return new ModelAndView(new RedirectView(topicUrl, false, false));
    } else {
      params.put("url", topicUrl);
      params.put("authorized", AuthUtil.isSessionAuthorized());

      return new ModelAndView("add-done-moderated", params);
    }
  }

  private static Poll preparePollPreview(AddTopicRequest form) {
    List<PollVariant> variants = new ArrayList<>(form.getPoll().length);

    for (String item : form.getPoll()) {
      if (!Strings.isNullOrEmpty(item)) {
        variants.add(new PollVariant(0, item));
      }
    }

    return Poll.apply(0, 0, form.isMultiSelect(), variants);
  }

  @RequestMapping("/add-section.jsp")
  public ModelAndView showForm(
          @RequestParam("section") int sectionId,
          @RequestParam(value="tag", required = false) String tag,
          HttpServletRequest request
  ) throws UserErrorException {
    Map<String, Object> params = new HashMap<>(prepareModel(null, AuthUtil.getCurrentUser(), sectionId, request));

    if (tag!=null) {
      TagName.checkTag(tag);
      params.put("tag", tag);
    }

    Section section = sectionService.getSection(sectionId);

    params.put("groups", groupDao.getGroups(section));

    return new ModelAndView("add-section", params);
  }

  @InitBinder
  public void initBinder(WebDataBinder binder) {
    binder.registerCustomEditor(Group.class, new PropertyEditorSupport() {
      @Override
      public void setAsText(String text) throws IllegalArgumentException {
        setValue(groupDao.getGroup(Integer.parseInt(text)));
      }

      @Override
      public String getAsText() {
        if (getValue()==null) {
          return null;
        } else {
          return Integer.toString(((Group) getValue()).getId());
        }
      }
    });

    binder.registerCustomEditor(User.class, new UserPropertyEditor(userService));
  }

  @InitBinder("form")
  public void requestValidator(WebDataBinder binder) {
    binder.setValidator(addTopicRequestValidator);

    binder.setBindingErrorProcessor(new ExceptionBindingErrorProcessor());
  }

  @ModelAttribute("modes")
  public Map<String, String> getModes() {
    Template tmpl = Template.getTemplate();

    return MessageTextService.postingModeSelector(AuthUtil.getCurrentUser(), tmpl.getFormatMode());
  }

  public static String getAddUrl(Section section, String tag) {
    UriComponentsBuilder builder = UriComponentsBuilder.fromPath("/add-section.jsp");
    builder.queryParam("section", section.getId());
    builder.queryParam("tag", tag);

    return builder.build().toUriString();
  }

  public static String getAddUrl(Section section) {
    UriComponentsBuilder builder = UriComponentsBuilder.fromPath("/add-section.jsp");
    builder.queryParam("section", section.getId());

    return builder.build().toUriString();
  }
}

File: src/main/scala/ru/org/linux/topic/AddTopicController.scala
/*
 * Copyright 1998-2026 Linux.org.ru
 *    Licensed under the Apache License, Version 2.0 (the "License");
 *    you may not use this file except in compliance with the License.
 *    You may obtain a copy of the License at
 *
 *        http://www.apache.org/licenses/LICENSE-2.0
 *
 *    Unless required by applicable law or agreed to in writing, software
 *    distributed under the License is distributed on an "AS IS" BASIS,
 *    WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
 *    See the License for the specific language governing permissions and
 *    limitations under the License.
 */
package ru.org.linux.topic

import com.google.common.base.Strings
import jakarta.servlet.ServletContext
import jakarta.servlet.http.HttpServletRequest
import org.apache.commons.io.IOUtils
import org.apache.pekko.actor.typed.ActorRef
import org.springframework.beans.factory.annotation.Qualifier
import org.springframework.http.HttpStatus
import org.springframework.security.crypto.password.PasswordEncoder
import org.springframework.stereotype.Controller
import org.springframework.validation.BindingResult
import org.springframework.web.bind.WebDataBinder
import org.springframework.web.bind.annotation.*
import org.springframework.web.servlet.ModelAndView
import org.springframework.web.servlet.view.RedirectView
import org.springframework.web.util.UriComponentsBuilder
import ru.org.linux.auth.*
import ru.org.linux.auth.AuthUtil.{MaybeAuthorized, MaybeAuthorizedCtx}
import ru.org.linux.csrf.{CSRFNoAuto, CSRFProtectionService}
import ru.org.linux.gallery.UploadedImagePreview
import ru.org.linux.group.GroupPermissionService.TopicLimitInfo
import ru.org.linux.group.{Group, GroupPermissionService, GroupService}
import ru.org.linux.msgbase.MessageText
import ru.org.linux.poll.{Poll, PollVariant}
import ru.org.linux.realtime.RealtimeEventHub
import ru.org.linux.rights.{AddTopicChecker, Permission, TopicPublishChecker}
import ru.org.linux.search.SearchQueueSender
import ru.org.linux.section.SectionController.NonTech
import ru.org.linux.section.{Section, SectionNotFoundException, SectionService}
import ru.org.linux.tag.TagService.tagRef
import ru.org.linux.tag.{TagName, TagService}
import ru.org.linux.user.{User, UserPropertyEditor, UserService}
import ru.org.linux.util.ExceptionBindingErrorProcessor
import ru.org.linux.util.markdown.MarkdownFormatter

import java.beans.PropertyEditorSupport
import java.nio.charset.StandardCharsets
import javax.annotation.Nullable
import javax.validation.Valid
import scala.beans.{BeanProperty, BooleanBeanProperty}
import scala.collection.mutable
import scala.jdk.CollectionConverters.*

@Controller
object AddTopicController:
  case class SectionChoice(
      @BeanProperty
      section: Section,
      @BeanProperty
      url: String,
      @BooleanBeanProperty
      postable: Boolean,
      @BeanProperty
      postScoreInfo: String)
  case class GroupChoice(
      @BeanProperty
      group: Group,
      @BeanProperty
      addUrl: String,
      @BooleanBeanProperty
      postable: Boolean,
      @BeanProperty
      postScoreInfo: String)

  private val MaxMessageLengthAnonymous = 8196
  private val MaxMessageLength = 65536

  private def preparePollPreview(form: AddTopicRequest): Poll =
    val variants = form.poll.filterNot(Strings.isNullOrEmpty).map(item => PollVariant(0, item))

    Poll(0, 0, form.multiSelect, variants.toVector)

  def getAddUrl(section: Section, tag: String): String =
    val builder = UriComponentsBuilder.fromPath("/add-section.jsp")
    builder.queryParam("section", section.id)
    builder.queryParam("tag", tag)
    builder.build.toUriString

  def getAddUrl(section: Section): String =
    val builder = UriComponentsBuilder.fromPath("/add-section.jsp")
    builder.queryParam("section", section.id)
    builder.build.toUriString

  def getAddUrl(group: Group): String = getAddUrl(group, null)

  def getAddUrl(
      group: Group,
      @Nullable
      tag: String,
      noinfo: Boolean = false): String =
    val builder = UriComponentsBuilder.fromPath("/add.jsp")
    builder.queryParam("group", group.id)
    if tag != null then
      builder.queryParam("tags", tag)
    if noinfo then
      builder.queryParam("noinfo", 1)
    builder.build.toUriString

@Controller
class AddTopicController(
                          searchQueueSender: SearchQueueSender,
                          captcha: CaptchaService,
                          sectionService: SectionService,
                          tagService: TagService,
                          userService: UserService,
                          prepareService: TopicPrepareService,
                          permissionService: GroupPermissionService,
                          addTopicRequestValidator: AddTopicRequestValidator,
                          topicService: TopicService,
                          @Qualifier("realtimeHubWS")
                          realtimeHubWS: ActorRef[RealtimeEventHub.Protocol],
                          renderService: MarkdownFormatter,
                          groupService: GroupService,
                          dupeProtector: FloodProtector,
                          servletContext: ServletContext,
                          addTopicChecker: AddTopicChecker,
                          topicPublishChecker: TopicPublishChecker,
                          passwordEncoder: PasswordEncoder):
  @RequestMapping(value = Array("/add.jsp"), method = Array(RequestMethod.GET))
  def add(
      @Valid @ModelAttribute("form")
      form: AddTopicRequest): ModelAndView =
    MaybeAuthorizedCtx {
      val group = form.group

      addTopicChecker.checkTopicPosting(group).checkOrThrow("Недостаточно прав для создания топика")

      val section = sectionService.getSection(form.group.sectionId)

      form.uploadedImages = new Array[String](permissionService.imageLimit(section))

      val topicLimitInfo = permissionService.topicLimitInfo(section)
      val topicPostingCheck = topicPublishChecker.checkPublish(group, topicLimitInfo)

      val params = prepareModelForm(Some(form.group), section, topicLimitInfo, topicPostingCheck)

      new ModelAndView("add", params.asJava)
    }

  private def prepareModelForm(group: Option[Group], section: Section, topicLimitInfo: TopicLimitInfo,
                               topicPostingCheck: Permission)
                              (using AnySession): Map[String, AnyRef] =
    prepareModel(group, section) +
      ("topicLimitInfo" -> topicLimitInfo) +
      ("topicPostingAllowed" -> Boolean.box(topicPostingCheck.permitted)) +
      ("topicPostingReason" -> topicPostingCheck.reason)

  private def prepareModel(group: Option[Group], section: Section)(using AnySession): Map[String, AnyRef] =
    val params = Map.newBuilder[String, AnyRef]

    val helpResource = servletContext.getResource("/help/new-topic-" + Section.getUrlName(section.id) + ".md")
    if helpResource != null then
      val helpRawText = IOUtils.toString(helpResource, StandardCharsets.UTF_8)
      val addInfo = renderService.renderToHtml(helpRawText, nofollow = false)
      params.addOne("addportal" -> addInfo)

    params.addOne("sectionId" -> Integer.valueOf(section.id))
    params.addOne("section" -> section)

    group.foreach { group =>
      params.addOne("group" -> group)
      params.addOne("showAllowAnonymous" -> Boolean.box(permissionService.enableAllowAnonymousCheckbox(group)))
      params.addOne("imagepost" -> Boolean.box(permissionService.isImagePostingAllowed(section)))
    }

    params.result()

  @RequestMapping(value = Array("/add.jsp"), method = Array(RequestMethod.POST)) @CSRFNoAuto
  def doAdd(
      request: HttpServletRequest,
      @Valid @ModelAttribute("form")
      form: AddTopicRequest,
      errors: BindingResult,
      @RequestAttribute("captchaRequired")
      captchaRequired: Boolean): ModelAndView =
    MaybeAuthorized { sessionUserOpt =>
      // не используем implicit, так как есть sessionUser и postingUser
      val group = form.group
      val section = sectionService.getSection(group.sectionId)


      if !form.isPreviewMode && !errors.hasErrors && captchaRequired then
        captcha.checkCaptcha(request, errors)

      val postingUser = AuthUtil.postingUser(sessionUserOpt, Option(form.nick), Option(form.password), errors, passwordEncoder, request)
      val user = postingUser.user

      val postingCheck = addTopicChecker.checkTopicPosting(group)(using postingUser)

      postingCheck.checkOrError(errors, "Недостаточно прав для создания топика")

      val topicLimitInfo = permissionService.topicLimitInfo(section)(using postingUser)
      val topicPostingCheck = topicPublishChecker.checkPublish(group, topicLimitInfo)(using postingUser)

      val params = prepareModelForm(Some(group), section, topicLimitInfo, topicPostingCheck)
        (using sessionUserOpt).to(mutable.HashMap)

      if !permissionService.enableAllowAnonymousCheckbox(group)(using postingUser) then
        form.allowAnonymous = true

      val message = MessageText(Strings.nullToEmpty(form.msg), sessionUserOpt.profile.formatMode)

      if !postingUser.authorized then
        if message.text.length > AddTopicController.MaxMessageLengthAnonymous then
          errors.rejectValue("msg", null, "Слишком большое сообщение")
      else if message.text.length > AddTopicController.MaxMessageLength then
        errors.rejectValue("msg", null, "Слишком большое сообщение")

      val imagePreviews =
        postingUser.opt match
          case Some(authorized) =>
            topicService.processUploads(form, group, errors)(using authorized)
          case None =>
            Seq.empty

      val poll: Option[Poll] =
        if section.isPollPostAllowed then
          Some(AddTopicController.preparePollPreview(form))
        else
          None

      val previewMsg: Topic = Topic.fromAddRequest(form, user, request.getRemoteAddr)

      val tagNames = TagName.parseAndSanitizeTags(form.tags)

      if !permissionService.canCreateTag(section)(using postingUser) then
        val newTags = tagService.getNewTags(tagNames)

        if newTags.nonEmpty then
          errors.rejectValue(
            "tags",
            null,
            "Вы не можете создавать новые теги (" + TagService.tagsToString(newTags) + ")")

      val preparedTopic = prepareService.prepareTopicPreview(
        previewMsg,
        tagNames.map(tagRef),
        poll,
        message,
        imagePreviews)(using postingUser)

      params.put("message", preparedTopic)

      val topicMenu = prepareService.getTopicMenu(preparedTopic, loadUserpics = true)(using sessionUserOpt)
      params.put("topicMenu", topicMenu)

      if !form.isPreviewMode && !errors.hasErrors then
        CSRFProtectionService.checkCSRF(request, errors)

      if !form.isPreviewMode && !errors.hasErrors then
        dupeProtector.checkRateLimit(FloodProtector.AddTopic, request.getRemoteAddr, postingUser.userOpt.orNull, errors)

      if !errors.hasErrors && !form.isDraftMode && !form.isPreviewMode then
        topicPostingCheck.checkOrError(errors)

      if !form.isPreviewMode && !errors.hasErrors then
        createNewTopic(
          request,
          form,
          group,
          params,
          section,
          user,
          message,
          imagePreviews,
          previewMsg)
      else
        new ModelAndView("add", params.asJava)
    }

  private def createNewTopic(
      request: HttpServletRequest,
      form: AddTopicRequest,
      group: Group,
      params: mutable.Map[String, AnyRef],
      section: Section,
      user: User,
      message: MessageText,
      images: Seq[UploadedImagePreview],
      previewMsg: Topic) =
    val (msgid, notifyUsers) = topicService.addMessage(
      request,
      form,
      message,
      group,
      user,
      images,
      previewMsg)

    if !previewMsg.draft then
      searchQueueSender.updateMessageOnly(msgid)
      RealtimeEventHub.notifyEvents(realtimeHubWS, notifyUsers)

    val topicUrl = previewMsg.withId(msgid).getLink

    if !section.premoderated || previewMsg.draft then
      new ModelAndView(new RedirectView(topicUrl, false, false))
    else
      params.put("url", topicUrl)

      new ModelAndView("add-done-moderated", params.asJava)

  @RequestMapping(path = Array("/add-section.jsp"), params = Array("section"))
  def showFormWithSection(
      @RequestParam("section")
      sectionId: Int,
      @RequestParam(value = "tag", required = false)
      tag: String): ModelAndView =
    MaybeAuthorizedCtx {
      val section = sectionService.getSection(sectionId)

      if tag != null then
        TagName.checkTag(tag)

      val groups = groupService.getGroups(section)

      if groups.size == 1 then
        val group = groups.head

        val postable = addTopicChecker.checkTopicPosting(group)

        if postable.permitted then
          new ModelAndView(new RedirectView(AddTopicController.getAddUrl(group, tag)))
        else
          val params = prepareModel(None, section).to(mutable.HashMap)

          params.put(
            "groups",
            Seq(
              AddTopicController.GroupChoice(
                group,
                AddTopicController.getAddUrl(group, tag, noinfo = true),
                false,
                postable.reason)).asJava)

          if tag != null then
            params.put("tag", tag)

          new ModelAndView("add-section", params.asJava)
      else
        val params = prepareModel(None, section).to(mutable.HashMap)

        val groupChoices = groups.map { group =>
          val postable = addTopicChecker.checkTopicPosting(group)

          AddTopicController.GroupChoice(
            group,
            AddTopicController.getAddUrl(group, tag, noinfo = true),
            postable.permitted,
            postable.reason)
        }

        if section.id == Section.Forum then
          val (other, tech) = groupChoices.partition(g => NonTech.contains(g.group.id))

          params.put("tech", tech.asJava)
          params.put("other", other.asJava)
        else
          params.put("groups", groupChoices.asJava)


        if tag != null then
          params.put("tag", tag)

        new ModelAndView("add-section", params.asJava)
    }

  @RequestMapping(path = Array("/add-section.jsp"), params = Array("!section"))
  def showFormAllSections(
      @RequestParam(value = "tag", required = false)
      tag: String): ModelAndView =
    MaybeAuthorizedCtx {
      val sectionList = sectionService
        .sections
        .map { section =>
          val groups = groupService.getGroups(section)

          val postable =
            if groups.size == 1 then
              addTopicChecker.checkTopicPosting(groups.head)
            else
              addTopicChecker.checkTopicPosting(section)

          val url =
            if groups.size == 1 then
              AddTopicController.getAddUrl(groups.head, tag)
            else
              val builder = UriComponentsBuilder.fromPath("/add-section.jsp")
              builder.queryParam("section", section.id)
              if tag != null then
                builder.queryParam("tag", tag)
              builder.build.toUriString

          AddTopicController.SectionChoice(section, url, postable.permitted, postable.reason)
        }

      val params = mutable.HashMap[String, AnyRef]("sectionList" -> sectionList.asJava)

      if tag != null then
        params.put("tag", tag)

      new ModelAndView("add-section", params.asJava)
    }

  @InitBinder
  def initBinder(binder: WebDataBinder): Unit =
    binder.registerCustomEditor(
      classOf[Group],
      new PropertyEditorSupport():
        override def setAsText(text: String): Unit = setValue(groupService.getGroup(text.toInt))

        override def getAsText: String =
          if getValue == null then
            null
          else
            getValue.asInstanceOf[Group].id.toString
    )

    binder.registerCustomEditor(classOf[User], new UserPropertyEditor(userService))

  @ExceptionHandler(Array(classOf[SectionNotFoundException])) @ResponseStatus(HttpStatus.NOT_FOUND)
  def handleNotFoundException: ModelAndView = new ModelAndView("errors/code404")

  @InitBinder(Array("form"))
  def requestValidator(binder: WebDataBinder): Unit =
    binder.setValidator(addTopicRequestValidator)
    binder.setBindingErrorProcessor(new ExceptionBindingErrorProcessor)
